File: Go/pi_sequential.go
//  A Go program to calculate π using quadrature as a sequential algorithm.
//
//  Copyright © 2010–2013 Russel Winder

package main

import (
	"./output"
	"time"
)

func main() {
	const n = 1000000000
	const delta = 1.0 / float64(n)
	startTime := time.Now()
	sum := float64(0.0)
	for i := 1; i <= n; i++ {
		x := (float64(i) - 0.5) * delta
		sum += 1.0 / (1.0 + x*x)
	}
	pi := 4.0 * delta * sum
	elapseTime := time.Now().Sub(startTime)
	output.OutS("Sequential", pi, n, elapseTime)
}

File: Haskell/pi_sequential.hs
--  Haskell implementation of π by Quadrature
--
--  Copyright © 2009–2011, 2013  Russel Winder

module Main where

import Output (out)

piIter :: Int -> Double -> Double -> Double
piIter 0 delta accumulator = 4.0 * delta * accumulator
piIter n delta accumulator = piIter (n - 1) delta (accumulator + 1.0 / (1.0 + x * x))
    where
      x = ((fromIntegral n) - 0.5) * delta

piQuad :: Int -> Double
piQuad n = piIter n (1.0 / (fromIntegral n)) 0.0

main :: IO()
main = do
  let n = 1000000000
  out "Sequential" (piQuad n) n
